prov/util/src/util_wait.c

```diff
@@ -424,3 +428,214 @@ int ofi_wait_fd_open(struct fid_fabric *fabric_fid, struct fi_wait_attr *attr,
 	free(wait);
 	return ret;
 }
+
+static void util_wait_yield_signal(struct util_wait *util_wait)
+{
+	(container_of(util_wait, struct util_wait_yield,
+			util_wait))->signal = 1;
+}
+
+static int util_wait_yield_try(struct util_wait *wait)
+{
+	struct ofi_wait_fid_entry *fid_entry;
+	struct util_wait_yield *wait_yield;
+	int ret = 0;
+
+	wait_yield = container_of(wait, struct util_wait_yield, util_wait);
+	fastlock_acquire(&wait_yield->lock);
+	wait_yield->signal = 0;
+	dlist_foreach_container(&wait_yield->fid_list, struct ofi_wait_fid_entry,
+				fid_entry, entry) {
+		ret = fid_entry->wait_try(fid_entry->fid);
+		if (ret != FI_SUCCESS)
+			goto out;
+	}
+
+	if (!ret && !wait_yield->signal)
+		ret = -FI_EAGAIN;
```

Comment:
I'm sleepy, and I don't understand this check.  The if statement doesn't seem needed.

Reply:
Looking at it again, the first part of the check isn't needed since we already break on error so I"ll take that out. But the second part is needed. That checks the signal and tells the yield_run call (which is what calls yield_try) to break if something came in.

Reply:
Well, we set the signal to 0 just above this, so this check is racy as well.  Maybe fold the wait_try call directly into the wait run function below.  We should only need to reset signal once per wait call.

Reply:
The locking is ugly but I think the version I just pushed should solve the race conditions. Let me know if there's a simpler way to do this that you can think of...

Reply:
Sorry, saw a mistake, repushed